README.md
# Overview

> ...chews through your system dependencies, spitting out incompatible versions.

When onboarding new team members, ensuring their computer has everything needed to work on the project can be painful. Verchew is a command-line program and embeddable Python script to check the versions of your project's system dependencies. Its only external dependency is any Python interpreter, which should already be installed on macOS and most Linux-based operating systems.

[![Travis CI](https://img.shields.io/travis/com/jacebrowning/verchew/main.svg?label=unix)](https://app.travis-ci.com/jacebrowning/verchew)
[![AppVeyor](https://img.shields.io/appveyor/ci/jacebrowning/verchew/main.svg?label=windows)](https://ci.appveyor.com/project/jacebrowning/verchew)
[![Coverage Status](https://img.shields.io/coveralls/jacebrowning/verchew/main.svg)](https://coveralls.io/r/jacebrowning/verchew)
[![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/jacebrowning/verchew.svg)](https://scrutinizer-ci.com/g/jacebrowning/verchew/?branch=main)
[![PyPI License](https://img.shields.io/pypi/l/verchew.svg)](https://pypi.org/project/verchew)
[![PyPI Version](https://img.shields.io/pypi/v/verchew.svg)](https://pypi.python.org/pypi/verchew)
[![PyPI Downloads](https://img.shields.io/pypi/dm/verchew.svg?color=orange)](https://pypistats.org/packages/verchew)

# Setup

## Requirements

- Python 2.7+ or Python 3.3+

## Installation

Install `verchew` globally with [pipx](https://pipxproject.github.io/pipx/installation/) (or pip):

```text
$ pipx install verchew
```

or add it to your [Poetry](https://poetry.eustace.io/) project:

```text
$ poetry add verchew
```

or embedded the script in your project using [this guide](https://verchew.readthedocs.io/en/latest/cli/vendoring/).

# Usage

Run `verchew --init` to generate a sample configuration file.

Update this file (`verchew.ini`) to include your project's system dependencies:

```ini
[Working Program]

cli = working-program
version = 1.2

[Newer Working Program]

cli = working-program
version =  4.1 || 4.2
message = Version 4.x is required to get the special features.

[Broken Program]

cli = broken-program
version = 1.2.3

[Optional Missing Program]

cli = missing-program
version = 1.2.3
optional = true

[Missing Program]

cli = missing-program
version = 1.2.3
```

Run `verchew` to see if you have the expected versions installed:

```text
$ verchew

Checking for Working Program...

$ working-program --version
1.2.3
✔ MATCHED: 1.2

Checking for Newer Working Program...

$ working-program --version
1.2.3
✘ EXPECTED: 4.1 || 4.2
䷉ MESSAGE: Version 4.x is required to get the special features.

Checking for Broken Program...

$ broken-program --version
An error occurred.
✘ EXPECTED: 1.2.3

Checking for Optional Missing Program...

$ missing-program --version
sh: command not found: missing-program
▴ EXPECTED (OPTIONAL): 1.2.3

Checking for Missing Program...

$ missing-program --version
sh: command not found: missing-program
✘ EXPECTED: 1.2.3

Results: ✔ ✘ ✘ ▴ ✘
```


CHANGELOG.md
# Release Notes

## 3.4.2 (2023-09-15)

- Fixed handling of `asdf` "not found" error messages.

## 3.4.1 (2023-06-13)

- Fixed handling of programs with no output.

## 3.4 (2023-06-03)

- Added support for matching version not in the first line of output.

## 3.3 (2022-04-18)

- Added a `verchew-wrapper` alongside the vendored script to support systems not following [PEP 394](https://peps.python.org/pep-0394/).

## 3.2.1 (2022-01-12)

- Fix handling of `optional=false` in configuration files.

## 3.2 (2022-01-07)

- Added support for passing multiple version arguments.

## 3.1 (2020-04-18)

- Added support for unicode characters in Python 3.8.
- Added `--vendor PATH` command to download the program for offline use.
- Updated `--quiet` output to include actual program versions.

## 3.0.2 (2020-01-06)

- Removed type annotations to continue supporting legacy Python.

## 3.0.1 (2019-12-13)

- Fixed missing program detection to only look for "not found" in the first line.

## 3.0 (2019-12-07)

- Removed `versions` in favor of `version` option.
- Removed `|` in favor of `||` for checking multiple versions.
- Added support for matching **any** version when `version` is not specified.

## 2.0.1 (2019-10-13)

- Restore compatibility layer to unofficially support legacy Python.

## 2.0 (2019-10-13)

- Deprecated `versions` in favor of `version` option.
- Deprecated `|` in favor of `||` for checking multiple versions.
- Stopped testing against Python 2, Python 3.4, and Python 3.5.

## 1.6.3 (2019-10-12)

- Fixed handling of programs that produce no output.

## 1.6.2 (2019-02-15)

- Fixed handling of programs that produce output with file paths.

## 1.6.1 (2019-02-14)

- Fixed handling of programs that produce no output.

## 1.6 (2019-02-08)

- Added `--quiet` option to suppress output when all programs match.

## 1.5 (2018-12-24)

- Updated `version` to only match at word boundaries.

## 1.4 (2018-03-06)

- Updated `cli_version_arg` to be blank for programs with no version argument.

## 1.3 (2018-02-08)

- Added support for checking for multiple versions of programs.

## 1.2 (2017-09-13)

- Added `optional = true` settings to downgrade errors to warnings.
- Added `message` setting to display optional message for missing programs.

## 1.1 (2017-05-17)

- Added `--init` command to generate a sample configuration file.

## 1.0 (2017-01-09)

- Initial stable release.

## 0.5 (2016-11-02)

- Added `--exit-code` option to return non-zero exit codes on failure.

## 0.4 (2016-10-20)

- Added support for custom version arguments.
- Fixed handling of internal versions matches.

## 0.3 (2016-10-18)

- Added support for detecting missing and broken programs.

## 0.2.1 (2016-10-18)

- Now stripping whitespace after calling `--version` on a program.

## 0.2 (2016-10-17)

- Added `--version` command.

## 0.1 (2016-10-17)

- Initial release.


verchew/script.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# The MIT License (MIT)
# Copyright © 2016, Jace Browning
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
# Source: https://github.com/jacebrowning/verchew
# Documentation: https://verchew.readthedocs.io
# Package: https://pypi.org/project/verchew


from __future__ import unicode_literals

import argparse
import logging
import os
import re
import sys
from collections import OrderedDict
from subprocess import PIPE, STDOUT, Popen

import subprocess
import threading

PY2 = sys.version_info[0] == 2

if PY2:
    import ConfigParser as configparser
    from urllib import urlretrieve
    from Queue import Queue
else:
    import configparser
    from urllib.request import urlretrieve
    from queue import Queue

__version__ = '3.4.1'

SCRIPT_URL = (
    "https://raw.githubusercontent.com/jacebrowning/verchew/main/verchew/script.py"
)
WRAPPER_URL = (
    "https://raw.githubusercontent.com/jacebrowning/verchew/main/verchew/wrapper.sh"
)

CONFIG_FILENAMES = ['verchew.ini', '.verchew.ini', '.verchewrc', '.verchew']

SAMPLE_CONFIG = """
[Python]

cli = python
version = Python 3.5 || Python 3.6

[Legacy Python]

cli = python2
version = Python 2.7

[virtualenv]

cli = virtualenv
version = 15
message = Only required with Python 2.

[Make]

cli = make
version = GNU Make
optional = true

""".strip()

STYLE = {
    "~": "✔",
    "?": "▴",
    "x": "✘",
    "#": "䷉",
}

COLOR = {
    "~": "\033[92m",  # green
    "?": "\033[93m",  # yellow
    "x": "\033[91m",  # red
    "#": "\033[96m",  # cyan
    None: "\033[0m",  # reset
}

QUIET = False

log = logging.getLogger(__name__)


def main():
    global QUIET

    args = parse_args()
    configure_logging(args.verbose)
    if args.quiet:
        QUIET = True

    log.debug("PWD: %s", os.getenv('PWD'))
    log.debug("PATH: %s", os.getenv('PATH'))

    if args.vendor:
        vendor_script(SCRIPT_URL, args.vendor)
        vendor_script(WRAPPER_URL, args.vendor + "-wrapper")
        sys.exit(0)

    path = find_config(args.root, generate=args.init)
    config = parse_config(path)

    if not check_dependencies(config) and args.exit_code:
        sys.exit(1)


def parse_args():
    parser = argparse.ArgumentParser(description="System dependency version checker.",)

    version = "%(prog)s v" + __version__
    parser.add_argument(
        '--version', action='version', version=version,
    )
    parser.add_argument(
        '-r', '--root', metavar='PATH', help="specify a custom project root directory"
    )
    parser.add_argument(
        '--exit-code',
        action='store_true',
        help="return a non-zero exit code on failure",
    )

    group_logging = parser.add_mutually_exclusive_group()
    group_logging.add_argument(
        '-v', '--verbose', action='count', default=0, help="enable verbose logging"
    )
    group_logging.add_argument(
        '-q', '--quiet', action='store_true', help="suppress all output on success"
    )

    group_commands = parser.add_argument_group('commands')
    group_commands.add_argument(
        '--init', action='store_true', help="generate a sample configuration file"
    )

    group_commands.add_argument(
        '--vendor', metavar='PATH', help="download the program for offline use"
    )

    args = parser.parse_args()

    return args


def configure_logging(count=0):
    if count == 0:
        level = logging.WARNING
    elif count == 1:
        level = logging.INFO
    else:
        level = logging.DEBUG

    logging.basicConfig(level=level, format="%(levelname)s: %(message)s")


def vendor_script(url, path):
    root = os.path.abspath(os.path.join(path, os.pardir))
    if not os.path.isdir(root):
        log.info("Creating directory %s", root)
        os.makedirs(root)

    log.info("Downloading %s to %s", url, path)
    urlretrieve(url, path)

    log.debug("Making %s executable", path)
    mode = os.stat(path).st_mode
    os.chmod(path, mode | 0o111)


def find_config(root=None, filenames=None, generate=False):
    root = root or os.getcwd()
    filenames = filenames or CONFIG_FILENAMES

    path = None
    log.info("Looking for config file in: %s", root)
    log.debug("Filename options: %s", ", ".join(filenames))
    for filename in os.listdir(root):
        if filename in filenames:
            path = os.path.join(root, filename)
            log.info("Found config file: %s", path)
            return path

    if generate:
        path = generate_config(root, filenames)
        return path

    msg = "No config file found in: {0}".format(root)
    raise RuntimeError(msg)


def generate_config(root=None, filenames=None):
    root = root or os.getcwd()
    filenames = filenames or CONFIG_FILENAMES

    path = os.path.join(root, filenames[0])

    log.info("Generating sample config: %s", path)
    with open(path, 'w') as config:
        config.write(SAMPLE_CONFIG + '\n')

    return path


def parse_config(path):
    data = OrderedDict()  # type: ignore

    log.info("Parsing config file: %s", path)
    config = configparser.ConfigParser()
    config.read(path)

    for section in config.sections():
        data[section] = OrderedDict()
        for name, value in config.items(section):
            data[section][name] = value

    for name in data:
        version = data[name].get('version') or ""
        data[name]['version'] = version
        data[name]['patterns'] = [v.strip() for v in version.split('||')]

        data[name]['optional'] = data[name].get(
            'optional', 'false'
        ).strip().lower() in ('true', 'yes', 'y', True)

    return data


def check_dependencies(config):
    success = []

    for name, settings in config.items():
        show("Checking for {0}...".format(name), head=True)
        output = get_version(settings['cli'], settings.get('cli_version_arg'))

        for pattern in settings['patterns']:
            if match_version(pattern, output):
                show(_("~") + " MATCHED: {0}".format(pattern or "<anything>"))
                success.append(_("~"))
                break
        else:
            if settings.get('optional'):
                show(_("?") + " EXPECTED (OPTIONAL): {0}".format(settings['version']))
                success.append(_("?"))
            else:
                if QUIET:
                    if "not found" in output:
                        actual = "Not found"
                    else:
                        actual = output.split('\n', maxsplit=1)[0].strip('.')
                    expected = settings['version'] or "<anything>"
                    print("{0}: {1}, EXPECTED: {2}".format(name, actual, expected))
                show(
                    _("x")
                    + " EXPECTED: {0}".format(settings['version'] or "<anything>")
                )
                success.append(_("x"))
            if settings.get('message'):
                show(_("#") + " MESSAGE: {0}".format(settings['message']))

    show("Results: " + " ".join(success), head=True)

    return _("x") not in success


def call_in_thread(args, queue):
    try:
        process = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        output, _ = process.communicate()
        queue.put((output, process.pid))
    except subprocess.CalledProcessError:
        queue.put(("sh: command not found: " + args[0], None))

def get_version(program, argument=None):
    queue = Queue()

    if argument is None:
        args = [program, '--version']
    elif argument:
        args = [program] + argument.split()
    else:
        args = [program]

    # First, try running with '--help'
    help_args = [program, '--help']
    help_thread = threading.Thread(target=call_in_thread, args=(help_args, queue))
    help_thread.start()
    help_thread.join(timeout=1)

    if help_thread.is_alive():
        # The '--help' command timed out, so don't try to get the version
        output = ""
    else:
        # The '--help' command succeeded, so try to get the version
        output = call(args)

    lines = output.splitlines()

    if lines:
        for line in lines:
            if any(char.isdigit() for char in line):
                show(line)
                break
        else:
            show(lines[0])
    else:
        show("<nothing>")

    return output


def match_version(pattern, output):
    lines = output.splitlines()
    if not lines or "not found" in lines[0]:
        return False

    regex = pattern.replace('.', r'\.') + r'(\b|/)'

    for line in lines:
        log.debug("Matching %s: %s", regex, line)
        match = re.match(regex, line)
        if match is None:
            log.debug("Matching %s: %s", regex, line)
            match = re.match(r'.*[^\d.]' + regex, line)
        if match:
            return True

    return False


def call(args):
    try:
        process = Popen(args, stdout=PIPE, stderr=STDOUT)
    except OSError:
        log.debug("Command not found: %s", args[0])
        output = "sh: command not found: {0}".format(args[0])
    else:
        raw = process.communicate()[0]
        output = raw.decode('utf-8').strip()
        log.debug("Command output: %r", output)

    return output


def show(text, start='', end='\n', head=False):
    """Python 2 and 3 compatible version of print."""
    if QUIET:
        return

    if head:
        start = '\n'
        end = '\n\n'

    if log.getEffectiveLevel() < logging.WARNING:
        log.info(text)
    else:
        formatted = start + text + end
        if PY2:
            formatted = formatted.encode('utf-8')
        sys.stdout.write(formatted)
        sys.stdout.flush()


def _(word, is_tty=None, supports_utf8=None, supports_ansi=None):
    """Format and colorize a word based on available encoding."""
    formatted = word

    if is_tty is None:
        is_tty = hasattr(sys.stdout, 'isatty') and sys.stdout.isatty()
    if supports_utf8 is None:
        supports_utf8 = str(sys.stdout.encoding).lower() == 'utf-8'
    if supports_ansi is None:
        supports_ansi = sys.platform != 'win32' or 'ANSICON' in os.environ

    style_support = supports_utf8
    color_support = is_tty and supports_ansi

    if style_support:
        formatted = STYLE.get(word, word)

    if color_support and COLOR.get(word):
        formatted = COLOR[word] + formatted + COLOR[None]

    return formatted


if __name__ == '__main__':  # pragma: no cover
    main()
